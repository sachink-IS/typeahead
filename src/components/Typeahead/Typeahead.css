.typeahead-basic{
  display: inline-block;
  text-align: left;
  color: gray;
  font-size: medium;
}

.input-style{
  width: 100%;
  color: gray;
  font-size: medium;
  position: relative;
}

.state-search-label{
  margin-bottom: 10px;
}

.caret{
  position: relative;
  right: 18px;
  top: 4px;
  cursor: pointer;
}

.SelectedState{
  display: inline-block;
  position: fixed;
  top: 370px;
  color: gray;
  font-size: medium;
}
.inputParent {
  display:flex;
}