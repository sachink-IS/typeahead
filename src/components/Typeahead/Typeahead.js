import React, { useState, useEffect, useRef } from 'react';
import axios from 'axios';
import DropDown from "../DropDown/DropDown";
import '@fortawesome/fontawesome-free/css/all.min.css';
import "./Typeahead.css";

/**
 * I have two approaches in mind while designing typeahead
 * 1. calling url after every change in input string
 *    pros: in future we can add pagination for the result when the result set is very huge
 *    cons: it consumes server attention everytime there is a change in input string
 * 
 * 2. calling the api only for the first time and filtering data locally for the rest of the times 
 *    pros: less network activity and server attention
 *    cons: will be difficult to show the data when data is very huge
 * 
 */

function usePrevious(state) {

  const ref = useRef();
  useEffect(() => {
    ref.current = state;
  });
  return ref.current;
}

export default function Typeahead(props) {
  const [aStates, setAStates] = useState([]);
  const [sStateName, setSStateName] = useState("");
  const [hideDropDown, setHideDropDown] = useState(true);
  const [sSelectedState, setSSelectedState] = useState("")
  const prevSearchText = usePrevious(sStateName);

  useEffect(() => {
    if (prevSearchText && prevSearchText.lenth < sStateName.length) {
      filterStatesLocally(sStateName);
    } else {
      getStates();
    }
  }, [sStateName])

  const filterStatesLocally = (value) => {
    // empty the selected state on backspace
    let filteredStates = aStates.filter(function (el) {
      let index = el.name.toLowerCase().indexOf(value.toLowerCase());
      return index > -1;
    })
    setAStates(filteredStates)
  }

  const updateStateName = (e) => {
    let enteredText = e.target.value.trim();
    let charOnlyRegex = /^[A-Za-z]+$/;
    // will not allow any other key to enter other than alphabets and space
    let validateEnteredKey = enteredText.match(charOnlyRegex) || enteredText === "";
    let sValue = (validateEnteredKey) ? enteredText : sStateName;
    setSSelectedState("");
    setSStateName(sValue);
    (sValue) ? setHideDropDown(false) : setHideDropDown(true);
  }

  const getUrl = () => {
    let baseUrl = props.url;
    let params = (sStateName) ? '?value=' + sStateName : '';
    return baseUrl + params;
  }

  const getStates = async () => {
    if(localStorage.getItem('states')){
      console.log("getting data from localstorage : ", JSON.parse(localStorage.getItem('states')));
      setAStates(JSON.parse(localStorage.getItem('states')));   
    }else{
      await axios({
        method: "get",
        url: getUrl()
      }).then((response) => {
        setAStates(response.data);
        localStorage.setItem("states", JSON.stringify(response.data));
      })  
    }
  }

  const selectState = (state) => {
    setSSelectedState(state.name);
    setSStateName(state.name);
    setHideDropDown(true);
  }

  const divStyle = {
    width: props.width || '250px',
    margin: props.margin || '10px'
  }

  const showIcon = (!sStateName && hideDropDown) ? 
  <i className="caret fas fa-angle-down" onClick={() => setHideDropDown(false)}></i> :
  <i className="caret fas fa-times" onClick={() => {
    setHideDropDown(true);
    setSSelectedState("");
    setSStateName("");
  }}></i>

  return (
    <div style={divStyle} className="typeahead-basic">
      <div className="state-search-label">Search for a state</div>
      <div className="inputParent">
        <input name="StateName"
          type="text"
          placeholder="Search State"
          className="input-style"
          value={sStateName}
          onChange={updateStateName}
        />
        {showIcon}
      </div>
      <DropDown
        states={aStates}
        searchText={sStateName}
        selectState={selectState}
        hideDropDown={hideDropDown}
      />
      <div className="SelectedState">
        Selected State : {sSelectedState}
      </div>
    </div>
  )
}
