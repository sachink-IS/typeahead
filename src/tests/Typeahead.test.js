import React from 'react';
import Typeahead from '../components/Typeahead/Typeahead';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

it('Typeahead component gets rendered ', () => {
  const component = renderer.create(
    <Typeahead
      width = "200px"
      margin = "10px"
      url = 'https://states.sachinkis.repl.co/'
    />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

describe("<Typeahead url = 'https://states.sachinkis.repl.co/'/>", function(){
  const wrapper = mount(
    <Typeahead
      url = 'https://states.sachinkis.repl.co/'
    />
  );
  let inputBox = wrapper.find("input[type='text']");

  it("should render input box", function () {
    expect(inputBox).toHaveLength(1);
  });

  it("should not render any dropdown items", function(){
    let dropdown = wrapper.find('.stateDom');
    expect(dropdown).toHaveLength(0);
  });
})