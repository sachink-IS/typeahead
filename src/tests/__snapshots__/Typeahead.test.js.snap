// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typeahead component gets rendered  1`] = `
<div
  className="typeahead-basic"
  style={
    Object {
      "margin": "10px",
      "width": "200px",
    }
  }
>
  <div
    className="state-search-label"
  >
    Search for a state
  </div>
  <div
    className="inputParent"
  >
    <input
      className="input-style"
      name="StateName"
      onChange={[Function]}
      placeholder="Search State"
      type="text"
      value=""
    />
    <i
      className="caret fas fa-angle-down"
      onClick={[Function]}
    />
  </div>
  <div
    className="dropdownOutline"
  />
  <div
    className="SelectedState"
  >
    Selected State : 
    
  </div>
</div>
`;
